src/core/core.datasetController.js

```diff
@@ -220,53 +297,105 @@ helpers.extend(DatasetController.prototype, {
 		return type && new type({
 			_ctx: me.chart.ctx,
 			_datasetIndex: me.index,
-			_index: index
+			_index: index,
+			_parsed: {}
 		});
 	},
 
+	/**
+	 * @private
+	 */
+	_dataCheck: function() {
+		var me = this;
+		var dataset = me.getDataset();
+		var data = dataset.data || (dataset.data = []);
+
+		// In order to correctly handle data addition/deletion animation (an thus simulate
+		// real-time charts), we need to monitor these data modifications and synchronize
+		// the internal meta data accordingly.
+
+		if (helpers.isObject(data)) {
+			// Object data is currently monitored for replacement only
+			if (me._objectData === data) {
+				return false;
+			}
+			me._data = convertObjectDataToArray(data);
+			me._objectData = data;
+		} else {
+			if (me._data === data && arraysEqual(data, me._dataCopy)) {
```

Comment:
may be a performance bottleneck here as the data array gets long

Reply:
Yeah. This is to detect `chart.data.datasets[0].data[5] = 12` type of changes.
We could stop supporting that and provide a `set` method instead. Or investigate using `Proxy` to detect the changes without going through the array.
Another option is to limit the `arraysEqual` to some small number of elements.